File: src/main/java/org/cubewhy/celestial/event/EventManager.java
/*
 * Celestial Launcher <me@lunarclient.top>
 * License under GPLv3
 * Do NOT remove this note if you want to copy this file.
 */

package org.cubewhy.celestial.event;

import lombok.Getter;
import org.jetbrains.annotations.NotNull;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

public class EventManager {

    private static final Map<Class<? extends Event>, ArrayList<EventData>> REGISTRY_MAP = new HashMap<>();


    @Getter
    private static final EventManager instance = new EventManager();


    private static void sortListValue(final Class<? extends Event> clazz) {

        final ArrayList<EventData> flexibleArray = new ArrayList<EventData>();

        for (final byte b : EventPriority.valueArray) {
            for (EventData methodData : EventManager.REGISTRY_MAP.get(clazz)) {
                if (methodData.priority == b) {
                    flexibleArray.add(methodData);
                }
            }
        }

        EventManager.REGISTRY_MAP.put(clazz, flexibleArray);

    }

    private static boolean isMethodBad(@NotNull final Method method) {
        return method.getParameterTypes().length != 1 || !method.isAnnotationPresent(EventTarget.class);
    }

    private static boolean isMethodBad(final Method method, final Class<? extends Event> clazz) {
        return isMethodBad(method) || method.getParameterTypes()[0].equals(clazz);
    }

    public static ArrayList<EventData> get(final Class<? extends Event> clazz) {
        return REGISTRY_MAP.get(clazz);
    }

    public static void cleanMap(final boolean removeOnlyEmptyValues) {

        final Iterator<Map.Entry<Class<? extends Event>, ArrayList<EventData>>> iterator = EventManager.REGISTRY_MAP.entrySet().iterator();

        while (iterator.hasNext()) {
            if (!removeOnlyEmptyValues || iterator.next().getValue().isEmpty()) {
                iterator.remove();
            }
        }
    }

    public static void unregister(final Object o, final Class<? extends Event> clazz) {

        if (REGISTRY_MAP.containsKey(clazz)) {
            for (final EventData methodData : REGISTRY_MAP.get(clazz)) {
                if (methodData.source.equals(o)) {
                    REGISTRY_MAP.get(clazz).remove(methodData);
                }
            }
        }

        cleanMap(true);

    }

    public static void unregister(final Object o) {

        for (ArrayList<EventData> flexibleArray : REGISTRY_MAP.values()) {

            for (int i = flexibleArray.size() - 1; i >= 0; i--) {

                if (flexibleArray.get(i).source.equals(o)) {
                    flexibleArray.remove(i);
                }

            }

        }

        cleanMap(true);

    }

    public static void register(@NotNull final Method method, final Object o) {

        final Class<?> clazz = method.getParameterTypes()[0];

        final EventData methodData = new EventData(o, method, method.getAnnotation(EventTarget.class).value());

        if (!methodData.target.isAccessible()) {
            methodData.target.setAccessible(true);
        }

        if (REGISTRY_MAP.containsKey(clazz)) {

            if (!REGISTRY_MAP.get(clazz).contains(methodData)) {
                REGISTRY_MAP.get(clazz).add(methodData);
                sortListValue((Class<? extends Event>) clazz);
            }

        } else {
            REGISTRY_MAP.put((Class<? extends Event>) clazz, new ArrayList<EventData>() {

                {
                    this.add(methodData);
                }

            });
        }

    }

    public static void register(@NotNull final Object o, final Class<? extends Event> clazz) {

        for (final Method method : o.getClass().getMethods()) {
            if (!isMethodBad(method, clazz)) {
                register(method, o);
            }
        }

    }

    public static void register(@NotNull Object o) {
        for (final Method method : o.getClass().getMethods()) {
            if (!isMethodBad(method)) {
                register(method, o);
            }
        }
    }

}

File: src/main/kotlin/org/cubewhy/celestial/event/EventManager.kt
/*
 * Celestial Launcher <me@lunarclient.top>
 * License under GPLv3
 * Do NOT remove this note if you want to copy this file.
 */
package org.cubewhy.celestial.event

import java.lang.reflect.Method

object EventManager {
    private val REGISTRY_MAP: MutableMap<Class<out Event?>, ArrayList<EventData>?> = HashMap()


    private fun sortListValue(clazz: Class<out Event?>) {
        val flexibleArray = ArrayList<EventData>()

        for (b in EventPriority.valueArray) {
            for (methodData in REGISTRY_MAP[clazz]!!) {
                if (methodData.priority == b) {
                    flexibleArray.add(methodData)
                }
            }
        }

        REGISTRY_MAP[clazz] = flexibleArray
    }

    private fun isMethodBad(method: Method): Boolean {
        return method.parameterTypes.size != 1 || !method.isAnnotationPresent(EventTarget::class.java)
    }

    private fun isMethodBad(method: Method, clazz: Class<out Event?>): Boolean {
        return isMethodBad(method) || method.parameterTypes[0] == clazz
    }

    fun get(clazz: Class<out Event?>): ArrayList<EventData>? {
        return REGISTRY_MAP[clazz]
    }

    fun cleanMap(removeOnlyEmptyValues: Boolean) {
        val iterator: MutableIterator<Map.Entry<Class<out Event?>, ArrayList<EventData>?>> =
            REGISTRY_MAP.entries.iterator()

        while (iterator.hasNext()) {
            if (!removeOnlyEmptyValues || iterator.next().value!!.isEmpty()) {
                iterator.remove()
            }
        }
    }

    fun unregister(o: Any, clazz: Class<out Event?>) {
        if (REGISTRY_MAP.containsKey(clazz)) {
            for (methodData in REGISTRY_MAP[clazz]!!) {
                if (methodData.source == o) {
                    REGISTRY_MAP[clazz]!!.remove(methodData)
                }
            }
        }

        cleanMap(true)
    }

    fun unregister(o: Any) {
        for (flexibleArray in REGISTRY_MAP.values) {
            for (i in flexibleArray!!.indices.reversed()) {
                if (flexibleArray[i].source == o) {
                    flexibleArray.removeAt(i)
                }
            }
        }

        cleanMap(true)
    }

    fun register(method: Method, o: Any?) {
        val clazz = method.parameterTypes[0]

        val methodData = EventData(
            o!!, method, method.getAnnotation(
                EventTarget::class.java
            ).value
        )

        if (!methodData.target.isAccessible) {
            methodData.target.isAccessible = true
        }

        if (REGISTRY_MAP.containsKey(clazz)) {
            if (!REGISTRY_MAP[clazz]!!.contains(methodData)) {
                REGISTRY_MAP[clazz]!!.add(methodData)
                sortListValue(clazz as Class<out Event?>)
            }
        } else {
            REGISTRY_MAP[clazz as Class<out Event?>] = object : ArrayList<EventData>() {
                init {
                    this.add(methodData)
                }
            }
        }
    }

    fun register(o: Any, clazz: Class<out Event?>) {
        for (method in o.javaClass.methods) {
            if (!isMethodBad(method, clazz)) {
                register(method, o)
            }
        }
    }

    @JvmStatic
    fun register(o: Any) {
        for (method in o.javaClass.methods) {
            if (!isMethodBad(method)) {
                register(method, o)
            }
        }
    }
}
